File: src/main/java/org/example/downloader/core/WebScrapers.java
package org.example.downloader.core;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.example.downloader.utils.JSoupDriver;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxDriverLogLevel;
import org.openqa.selenium.firefox.FirefoxOptions;

import java.util.logging.Level;

public class WebScrapers {
    private static JSoupDriver jSoupDriver;
    private static FirefoxDriver firefoxDriver;

    /**
     * Returns the WebDriver responsible for handling non js webscraping.
     *
     * @return the JSoupDriver
     */
    public static JSoupDriver noJavaScript() {
        return jSoupDriver;
    }

    /**
     * Returns the WebDriver responsible for handling JavaScript webscraping.
     *
     * @return the FirefoxDriver
     */
    public static WebDriver javaScript() {
        return firefoxDriver;
    }

    /**
     * If one of the WebDrivers happens to be closed, you can reinitialize them with this method.
     */
    public static void reInitialize() {
        if (firefoxDriver.toString().contains("null")) {
            FirefoxOptions options = new FirefoxOptions().setHeadless(true).setLogLevel(FirefoxDriverLogLevel.FATAL).addArguments("--log", "fatal");
            options.setCapability("marionette", true);
            (firefoxDriver = new FirefoxDriver(options)).setLogLevel(Level.OFF);
        }
        if (jSoupDriver.toString().contains("null")) {
            jSoupDriver = new JSoupDriver();
        }
    }

    /**
     * Closes both WebDrivers to release resources.
     */
    public static void close() {
        firefoxDriver.quit();
        jSoupDriver.quit();
    }

    /**
     * Initializes the WebScraper for handling (non-)JavaScript websites.
     */
    public void initialize() {
        java.util.logging.Logger.getLogger("org.openqa.selenium").setLevel(Level.SEVERE);

        WebDriverManager.firefoxdriver().setup();

        FirefoxOptions options = new FirefoxOptions().setHeadless(true).setLogLevel(FirefoxDriverLogLevel.FATAL).addArguments("--log", "fatal");
        options.setCapability("marionette", true);
        (firefoxDriver = new FirefoxDriver(options)).setLogLevel(Level.OFF);
        jSoupDriver = new JSoupDriver();
    }
}

File: server/src/main/kotlin/eu/timerertim/downlomatic/util/WebScrapers.kt
package eu.timerertim.downlomatic.util

import io.github.bonigarcia.wdm.WebDriverManager
import org.openqa.selenium.WebDriver
import org.openqa.selenium.firefox.FirefoxDriver
import org.openqa.selenium.firefox.FirefoxDriverLogLevel
import org.openqa.selenium.firefox.FirefoxOptions
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Provides webscrapers for usage. Has to be initialized at least once in order to work.
 */
object WebScrapers {
    private lateinit var jSoupDriver: JSoupDriver
    private lateinit var firefoxDriver: FirefoxDriver

    /**
     * Returns the WebDriver responsible for handling non js webscraping.
     *
     * @return the JSoupDriver
     */
    @JvmStatic
    fun noJavaScript(): JSoupDriver {
        if (this::jSoupDriver.isInitialized.not()) {
            initializeJSoupDriver()
        }
        return jSoupDriver
    }

    /**
     * Returns the WebDriver responsible for handling JavaScript webscraping.
     *
     * @return the FirefoxDriver
     */
    @JvmStatic
    fun javaScript(): WebDriver {
        if (this::firefoxDriver.isInitialized.not()) {
            initializeFirefoxDriver()
        }
        return firefoxDriver
    }

    /**
     * If one of the WebDrivers happens to be closed, you can reinitialize them with this method.
     */
    fun reInitialize() {
        if (firefoxDriver.toString().contains("null")) {
            initializeFirefoxDriver()
        }
        if (jSoupDriver.toString().contains("null")) {
            initializeJSoupDriver()
        }
    }

    /**
     * Closes both WebDrivers to release resources.
     */
    @JvmStatic
    fun close() {
        if (this::firefoxDriver.isInitialized) firefoxDriver.quit()
        if (this::jSoupDriver.isInitialized) jSoupDriver.quit()
    }

    /**
     * Initializes the WebScraper for handling (non-)JavaScript websites.
     */
    @JvmStatic
    fun initialize() {
        initializeFirefoxDriver()
        initializeJSoupDriver()
    }

    private fun initializeFirefoxDriver() {
        Logger.getLogger("org.apache.hc.client5.http").level = Level.OFF
        Logger.getLogger("org.apache.logging").level = Level.OFF
        WebDriverManager.firefoxdriver().setup()

        Logger.getLogger("org.openqa.selenium").level = Level.OFF
        val options =
            FirefoxOptions().setHeadless(true).setLogLevel(FirefoxDriverLogLevel.FATAL).addArguments("--log", "fatal")
        options.setCapability("marionette", true)
        FirefoxDriver(options).also { firefoxDriver = it }.setLogLevel(Level.OFF)
    }

    private fun initializeJSoupDriver() {
        jSoupDriver = JSoupDriver()
    }
}